src/System.Private.CoreLib/src/System/Utf8String.Construction.cs

```diff
@@ -159,47 +233,437 @@ private unsafe Utf8String Ctor(char* value)
+
+            long totalUtf8BytesRequired = (uint)value.Length + utf8CodeUnitCountAdjustment;
+            if (totalUtf8BytesRequired > int.MaxValue)
+            {
+                totalUtf8BytesRequired = int.MaxValue;
+            }
+
+            // We can get away with FastAllocateSkipZeroInit here because we're not going to return the
+            // new Utf8String instance to the caller if we don't overwrite every byte of the buffer.
+
+            newString = FastAllocateSkipZeroInit((int)totalUtf8BytesRequired);
+
+            // Now transcode the UTF-16 input into the newly allocated Utf8String's buffer. We can't call the
+            // "skip validation" transcoder because the caller could've mutated the input buffer between the
+            // initial counting step and the transcoding step below.
+
+            status = Utf8.FromUtf16(value, newString.DangerousGetMutableSpan(), out _, out int bytesWritten, replaceInvalidSequences: false);
+            if (status != OperationStatus.Done || bytesWritten != newString.Length)
+            {
+                // Did somebody mutate our input buffer? Shouldn't be any other way this could happen.
+
+                return null;
+            }
+
+            return newString;
+        }
+
+        /// <summary>
+        /// Creates a new <see cref="Utf8String"/> instance, allowing the provided delegate to populate the
+        /// instance data of the returned object.
+        /// </summary>
+        /// <typeparam name="TState">Type of the state object provided to <paramref name="action"/>.</typeparam>
+        /// <param name="length">The length, in bytes, of the <see cref="Utf8String"/> instance to create.</param>
+        /// <param name="state">The state object to provide to <paramref name="action"/>.</param>
+        /// <param name="action">The callback which will be invoked to populate the returned <see cref="Utf8String"/>.</param>
+        /// <exception cref="ArgumentException">
+        /// Thrown if <paramref name="action"/> populates the buffer with ill-formed UTF-8 data.
+        /// </exception>
+        /// <remarks>
+        /// The runtime will perform UTF-8 validation over the contents provided by the <paramref name="action"/> delegate.
+        /// If an invalid UTF-8 subsequence is detected, an exception is thrown.
+        /// </remarks>
+        public static Utf8String Create<TState>(int length, TState state, SpanAction<byte, TState> action)
+        {
+            if (length < 0)
+            {
+                ThrowHelper.ThrowLengthArgumentOutOfRange_ArgumentOutOfRange_NeedNonNegNum();
+            }
+
+            if (action is null)
+            {
+                ThrowHelper.ThrowArgumentNullException(ExceptionArgument.action);
+            }
+
+            // Create and populate the Utf8String instance.
+            // Can't use FastAllocateSkipZeroInit here because we're handing the raw buffer to user code.
```

Comment:
Move the comment below the `length == 0`-check?

Same below in `CreateRelaxed`, and `UnsafeCreateWithoutValidation`.